File: 056-merge-intervals/merge-intervals.py
# -*- coding:utf-8 -*-


# Given a collection of intervals, merge all overlapping intervals.
#
#
# For example,
# Given [1,3],[2,6],[8,10],[15,18],
# return [1,6],[8,10],[15,18].


# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: List[Interval]
        """
        if not intervals:
            return []
        
        intervals.sort(key = lambda x: x.start, reverse=True)
        stack = []
        tmp = intervals.pop()
        while intervals:
            x = intervals.pop()
            if tmp.end >= x.start:
                tmp.end = max(tmp.end, x.end)
            else:
                stack.append(tmp)
                tmp = x
        stack.append(tmp)
        return stack
            

File: 056-merge-intervals/merge-intervals.cpp
// Given a collection of intervals, merge all overlapping intervals.
//
// Example 1:
//
//
// Input: [[1,3],[2,6],[8,10],[15,18]]
// Output: [[1,6],[8,10],[15,18]]
// Explanation: Since intervals [1,3] and [2,6] overlaps, merge them into [1,6].
//
//
// Example 2:
//
//
// Input: [[1,4],[4,5]]
// Output: [[1,5]]
// Explanation: Intervals [1,4] and [4,5] are considered overlapping.
//


/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    vector<Interval> merge(vector<Interval>& intervals) {
        int n = intervals.size();
        if (!n) return {};
        sort(intervals.begin(), intervals.end(), [](Interval& a, Interval& b){return a.start < b.start;});
        vector<Interval> ans;
        ans.emplace_back(intervals[0]);
        for (int i = 1; i < n; i++) {
            int ts = intervals[i].start, te = intervals[i].end;
            if (ans.back().end < ts) {
                ans.emplace_back(intervals[i]);
            } else {
                ans.back().end = max(ans.back().end, te);
            }
        }
        return ans;
    }
};
